### ISSUE TITLE
Cross compilation from Linux with mingw fails in cmake module

### ISSUE THREAD

**Issue by rezahousseini** [mesonbuild/meson#7105]
When cross compiling a cmake project with the following cross file

```
[binaries]
c = '/usr/bin/x86_64-w64-mingw32.static-gcc'
cpp = '/usr/bin/x86_64-w64-mingw32.static-g++'
ld = '/usr/bin/x86_64-w64-mingw32.static-g++'
addr2line = '/usr/bin/x86_64-w64-mingw32.static-addr2line'
ar = '/usr/bin/x86_64-w64-mingw32.static-ar'
elfedit = '/usr/bin/x86_64-w64-mingw32.static-elfedit'
objcopy = '/usr/bin/x86_64-w64-mingw32.static-objcopy'
objdump = '/usr/bin/x86_64-w64-mingw32.static-objdump'
ranlib = '/usr/bin/x86_64-w64-mingw32.static-ranlib'
readelf = '/usr/bin/x86_64-w64-mingw32.static-readelf'
strip = '/usr/bin/x86_64-w64-mingw32.static-strip'
pkgconfig = '/usr/bin/x86_64-w64-mingw32.static-pkg-config'
windres = '/usr/bin/x86_64-w64-mingw32.static-windres'
cmake = '/usr/bin/x86_64-w64-mingw32.static-cmake'

[properties]
# Directory that contains 'bin', 'lib', etc
root = '/usr/x86_64-w64-mingw32.static'

[host_machine]
system = 'windows'
cpu_family = 'x86_64'
cpu = 'x86_64'
endian = 'little'
```

I get the following error for a cmake project (`meson-log.txt`):

```
Executing subproject di method cmake 

CMake binary for MachineChoice.HOST is cached.

Configuring the build directory with CMake version 3.15.4
Running: -DCMAKE_CXX_COMPILER=/usr/bin/x86_64-w64-mingw32.static-g++ -G Ninja -DCMAKE_INSTALL_PREFIX=/usr/local
  - build directory:          /usr/project/build/subprojects/di/__CMake_build
  - source directory:         /usr/project/subprojects/di
  - trace args:               --trace --trace-expand --no-warn-unused-cli
  - preload file:             /usr/local/lib/python3.6/dist-packages/meson-0.54.999-py3.6.egg/mesonbuild/cmake/data/preload.cmake
  - disabled policy warnings: [CMP0025, CMP0047, CMP0056, CMP0060, CMP0065, CMP0066, CMP0067, CMP0082, CMP0089]

== Using MXE wrapper: /usr/bin/x86_64-w64-mingw32.static-cmake
     - cmake version 3.15.4
     - warnings for unused CMAKE_POLICY_DEFAULT variables can be ignored
== Using MXE toolchain: /usr/x86_64-w64-mingw32.static/share/cmake/mxe-conf.cmake
== Using MXE runresult: /usr/share/cmake/modules/TryRunResults.cmake
== Adding "-DCMAKE_BUILD_TYPE=Release"
Running with trace output on.
Running with expanded trace output on.
Not searching for unused variables given on the command line.
loading initial cache file /usr/share/cmake/modules/TryRunResults.cmake
-- The CXX compiler identification is GNU 9.3.0
-- Check for working CXX compiler: /usr/x86_64-pc-linux-gnu/bin/x86_64-w64-mingw32.static-g++
-- Check for working CXX compiler: /usr/x86_64-pc-linux-gnu/bin/x86_64-w64-mingw32.static-g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
WARNING: CMake Warning (dev) at /usr/local/lib/python3.6/dist-packages/meson-0.54.999-py3.6.egg/mesonbuild/cmake/data/preload.cmake:46 (if):
WARNING:   Policy CMP0054 is not set: Only interpret if() arguments as variables or
WARNING:   keywords when unquoted.  Run "cmake --help-policy CMP0054" for policy
WARNING:   details.  Use the cmake_policy command to set the policy and suppress this
WARNING:   warning.
WARNING: 
WARNING:   Quoted variables like "PROPERTIES" will no longer be dereferenced when the
WARNING:   policy is set to NEW.  Since the policy is not set the OLD behavior will be
WARNING:   used.
WARNING: Call Stack (most recent call first):
WARNING:   test/CMakeLists.txt:67 (set_source_files_properties)
WARNING: This warning is for project developers.  Use -Wno-dev to suppress it.
WARNING: 
-- Configuring done
WARNING: You have changed variables that require your cache to be deleted.
WARNING: Configure will be re-run and you may have to reset some variables.
WARNING: The following variables have changed:
WARNING: CMAKE_CXX_COMPILER= /usr/bin/x86_64-w64-mingw32.static-g++
WARNING: 
-- The CXX compiler identification is GNU 9.3.0
-- Check for working CXX compiler: /usr/bin/x86_64-w64-mingw32.static-g++
WARNING: CMake Warning (dev) at /usr/x86_64-pc-linux-gnu/share/cmake-3.15/Modules/CMakeTestCXXCompiler.cmake:37 (try_compile):
WARNING:   Policy CMP0066 is not set: Honor per-config flags in try_compile()
WARNING:   source-file signature.  Run "cmake --help-policy CMP0066" for policy
WARNING:   details.  Use the cmake_policy command to set the policy and suppress this
WARNING:   warning.
WARNING: 
WARNING:   For compatibility with older versions of CMake, try_compile is not honoring
WARNING:   caller config-specific compiler flags (e.g.  CMAKE_C_FLAGS_DEBUG) in the
WARNING:   test project.
WARNING: Call Stack (most recent call first):
WARNING:   CMakeLists.txt:8 (project)
WARNING: This warning is for project developers.  Use -Wno-dev to suppress it.
WARNING: 
WARNING: CMake Warning (dev) at /usr/x86_64-pc-linux-gnu/share/cmake-3.15/Modules/CMakeTestCXXCompiler.cmake:37 (try_compile):
WARNING:   Policy CMP0056 is not set: Honor link flags in try_compile() source-file
WARNING:   signature.  Run "cmake --help-policy CMP0056" for policy details.  Use the
WARNING:   cmake_policy command to set the policy and suppress this warning.
WARNING: 
WARNING:   For compatibility with older versions of CMake, try_compile is not honoring
WARNING:   caller link flags (e.g.  CMAKE_EXE_LINKER_FLAGS) in the test project.
WARNING: Call Stack (most recent call first):
WARNING:   CMakeLists.txt:8 (project)
WARNING: This warning is for project developers.  Use -Wno-dev to suppress it.
WARNING: 
-- Check for working CXX compiler: /usr/bin/x86_64-w64-mingw32.static-g++ -- broken
-- Configuring incomplete, errors occurred!
See also "/usr/project/build/subprojects/di/__CMake_build/CMakeFiles/CMakeOutput.log".
See also "/usr/project/build/subprojects/di/__CMake_build/CMakeFiles/CMakeError.log".

CMake configuration: FAILED

meson.build:15:0: ERROR: Failed to configure the CMake subproject
```
It seems that something is changing variables in the CMakeCache which triggers another compiler identification but this time without the correct toolchain file (detects GNU instead of Windows).

A workaround is to supply a toolchain file to cmake:

```
# the name of the target operating system
SET(CMAKE_SYSTEM_NAME Windows)

# which compilers to use for C and C++
SET(CMAKE_C_COMPILER x86_64-w64-mingw32.static-gcc)
SET(CMAKE_CXX_COMPILER x86_64-w64-mingw32.static-g++)
SET(CMAKE_RC_COMPILER x86_64-w64-mingw32.static-windres)

# here is the target environment located
SET(CMAKE_FIND_ROOT_PATH  /usr/x86_64-w64-mingw32.static /usr/project/ )

# adjust the default behaviour of the FIND_XXX() commands:
# search headers and libraries in the target environment, search
# programs in the host environment
set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
```

via the `cmake_option`:

```
di_proj = cmake.subproject('di', cmake_options: '-DCMAKE_TOOLCHAIN_FILE=/usr/project/windows-mingw-static.cmake')
```

this seem to work but is kind of ugly

**Comment by mensinda** [comment]
Cross-compilation with CMake subprojects is not supported yet. However, this will be resolved at some point.

**Comment by mensinda** [comment]
Implemented in #7816
